sound/soc/sof/ipc4.c

```diff
@@ -435,6 +438,11 @@ static int sof_ipc4_set_get_data(struct snd_sof_dev *sdev, void *data,
 
 	tx.extension |= SOF_IPC4_MOD_EXT_MSG_FIRST_BLOCK(1);
 
+	/* ensure the DSP is in D0i0 before sending IPC */
+	ret = snd_sof_dsp_set_power_state(sdev, &target_state);
+	if (ret < 0)
+		return ret;
+
```

Comment:
humm, didn't we have something in the past where we wouldn't access the data and wake-up the DSP if it was in D3?

Not sure what would happen if userspace plays with a kcontrol, is this really needed to wake-up the DSP?

Clearly not following this PR.

Reply:
@plbossart there's no waking up here. With deep buffer pipeline active, we could be in d0i3 and if we invoke set_get_data, we have to make sure we're in d0i0 first.

Reply:
well, exactly, wasn't it agreed that the IPC would trigger a move back to D0i0, so why would we need to do this again?

Reply:
Because this ops sends ipc without going through the check for that which is only done in sof_ipc4_tx_msg() and set_get_data() doesn't use that function. So we need to make sure we do it here as well.

Reply:
> Because this ops sends ipc without going through the check for that which is only done in sof_ipc4_tx_msg() and set_get_data() doesn't use that function. So we need to make sure we do it here as well.

I was wondering why only sof_ipc4_set_get_data need to be in D0i0 but not all IPC message. This comment answered my question.

Reply:
Fwiw, the IPC3 implementation is identical: `sof_ipc3_set_get_data()` uses `ipc3_tx_msg_unlocked()` and the set power state is in `sof_ipc3_tx_msg()`, which is not in that execution path.
There it is good, here it is not?

Reply:
@ujfalusi in IPC3, we do not send the set-get_data() IPC just before suspend. It should be fixed there too but we just dont see the problem

Reply:
> @ujfalusi in IPC3, we do not send the set-get_data() IPC just before suspend. It should be fixed there too but we just dont see the problem

We had relatively frequent suspend/resume issues all across the board since forever. They just became much more frequent recently.

Granted, some happen even without audio; suspend/resume is always broken by the weakest link in the chain.

@ranj063 will you submit the IPC3 fix?

Reply:
@marc-hb no I wont be fixing this for IPC3. It is not a sequence that is used with IPC3 i.e set_get_data() while the DSP is in D0i3. So there would be no way for me to test it.

Reply:
Sorry I misunderstood "It should be fixed there too but we just dont see the problem"